tsl/src/bgw_policy/continuous_aggregate_api.c

```diff
@@ -183,48 +211,211 @@ convert_interval_arg(Oid dim_type, Datum interval, Oid *interval_type, const cha
+
+	/* Widen dayfraction to 128 bits */
+	span = int64_to_int128(dayfraction);
+
+	/* Scale up days to microseconds, forming a 128-bit product */
+	int128_add_int64_mul_int64(&span, days, USECS_PER_DAY);
+
+	return span;
+}
+
+static int64
+interval_to_int64(Datum interval, Oid type)
+{
+	int64 result;
+
+	switch (type)
 	{
-		switch (interval_type)
+		case INT2OID:
+			result = DatumGetInt16(interval);
+			break;
+		case INT4OID:
+			result = DatumGetInt32(interval);
+			break;
+		case INT8OID:
+			result = DatumGetInt64(interval);
+			break;
+		case INTERVALOID:
 		{
-			case INT2OID:
-			{
-				if (DatumGetInt16(start_offset) <= DatumGetInt16(end_offset))
-					valid = false;
-				break;
-			}
-			case INT4OID:
-			{
-				if (DatumGetInt32(start_offset) <= DatumGetInt32(end_offset))
-					valid = false;
-				break;
-			}
-			case INT8OID:
-			{
-				if (DatumGetInt64(start_offset) <= DatumGetInt64(end_offset))
-					valid = false;
-				break;
-			}
+			const int64 max = ts_time_get_max(TIMESTAMPTZOID);
+			const int64 min = ts_time_get_min(TIMESTAMPTZOID);
+			INT128 bigres = interval_to_int128(DatumGetIntervalP(interval));
+
+			if (int128_compare(bigres, int64_to_int128(max)) >= 0)
+				result = max;
+			else if (int128_compare(bigres, int64_to_int128(min)) <= 0)
+				result = min;
+			else
+				result = int128_to_int64(bigres);
+			break;
 		}
+		default:
+			pg_unreachable();
 	}
+
+	return result;
+}
+
+/*
+ * Enforce that a policy has a refresh window of at least two buckets to
+ * ensure we materialize at least one bucket each run.
+ *
+ * Why two buckets? Note that the policy probably won't execute at at time
+ * that exactly aligns with a bucket boundary, so a window of one bucket
+ * might not cover a full bucket that we want to materialize:
+ *
+ * Refresh window:                   [-----)
+ * Materialized buckets:   |-----|-----|-----|
+ */
+static void
+validate_window_size(const ContinuousAgg *cagg, const CaggPolicyConfig *config)
+{
+	int64 start_offset;
+	int64 end_offset;
+
+	if (config->offset_isnull[OFFSET_START])
+		start_offset = ts_time_get_max(cagg->partition_type);
 	else
-	{
-		Assert(interval_type == INTERVALOID);
-		valid = DatumGetBool(DirectFunctionCall2(interval_gt, start_offset, end_offset));
-	}
-	if (!valid)
-	{
+		start_offset =
+			interval_to_int64(config->offset[OFFSET_START], config->offset_type[OFFSET_START]);
+
+	if (config->offset_isnull[OFFSET_END])
+		end_offset = ts_time_get_min(cagg->partition_type);
+	else
+		end_offset = interval_to_int64(config->offset[OFFSET_END], config->offset_type[OFFSET_END]);
+
+	if (ts_time_saturating_add(end_offset, cagg->data.bucket_width * 2, INT8OID) > start_offset)
 		ereport(ERROR,
 				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
-				 errmsg("start interval should be greater than end interval")));
+				 errmsg("policy refresh window too small"),
+				 errhint("The refresh window must be at least "
+						 "two buckets in the valid time range.")));
```

Comment:
IMO, errors here should be intelligible for users, not formatted for debugging. Such messages should be reserved for DEBUG-level output.

There are several issues with printing the suggested parameters here. For instance, the bucket width is specified in "internal" time and it isn't obvious how to present this in a useful way that doesn't read like just a large int64 number. The same goes for offset parameters. The user input, e.g., `start_offset=>'1 month'` but gets back an unintelligible int64 number.

That said, I am not opposed to well-formatted messages with additional information that follow style guide lines.

Reply:
Tweaked the error to include at least the desired bucket size. The messages are visible in the test output.

Reply:
@mkindahl Your example isn't correct since not all the the time types are int64 (we also support other integers and data/timestamp types). Further, the offsets are not time values, they are intervals (e.g. `INTERVALOID`). 

That said, if you check the code and test output you will find that I tweaked the error messages.

Reply:
It was just an example, I didn't intend to make a complete implementation.